googlecloud-plugins/src/main/java/co/cask/hydrator/plugin/sink/GCSBatchSink.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.hydrator.plugin.sink;
+
+import co.cask.cdap.api.annotation.Description;
+import co.cask.cdap.api.annotation.Macro;
+import co.cask.cdap.api.annotation.Name;
+import co.cask.cdap.api.data.batch.Output;
+import co.cask.cdap.api.data.batch.OutputFormatProvider;
+import co.cask.cdap.api.data.format.StructuredRecord;
+import co.cask.cdap.etl.api.batch.BatchSinkContext;
+import co.cask.hydrator.common.ReferenceBatchSink;
+import co.cask.hydrator.common.ReferencePluginConfig;
+import co.cask.hydrator.common.batch.sink.SinkOutputFormatProvider;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.reflect.TypeToken;
+import com.google.gson.Gson;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.reflect.Type;
+import java.util.HashMap;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * {@link GCSBatchSink} that stores the data to Google Cloud Storage Bucket.
+ * @param <KEY_OUT> the type of key the sink outputs
+ * @param <VAL_OUT> the type of value the sink outputs
+ */
+
+public abstract class GCSBatchSink<KEY_OUT, VAL_OUT> extends ReferenceBatchSink<StructuredRecord, KEY_OUT, VAL_OUT> {
+  public static final String BUCKET_DES = "GCS Bucket to use to store the data";
+  public static final String PROJECT_ID_DES = "Google Cloud Project ID with access to configured GCS buckets";
+  public static final String SERVICE_KEY_FILE_DES = "The Json_Key_File certificate file of the " +
+    "service account used for GCS access";
+  private static final String FILESYSTEM_PROPERTIES_DESCRIPTION = "A JSON string representing a map of properties " +
+    "needed for the distributed file system.";
+  private static final Gson GSON = new Gson();
+
+  private static final Type MAP_STRING_STRING_TYPE = new TypeToken<Map<String, String>>() { }.getType();
+
+  private final GCSSinkConfig config;
+
+  protected GCSBatchSink(GCSSinkConfig config) {
+    super(config);
+    this.config = config;
+    if (!this.config.containsMacro("fileSystemProperties") && !this.config.containsMacro("Bucket_Key") &&
+      !this.config.containsMacro("Project_Id")) {
+      this.config.fileSystemProperties = updateFileSystemProperties(this.config.fileSystemProperties,
```

Comment:
If these fields contained macros, then `config.fileSystemProperties` will not be set in the constructor, you will have to set it in `prepareRun` in that case.